### ISSUE TITLE
ecdaq_case calibration/execution problem

### ISSUE THREAD

**Issue by miguelaferreira** [cupertino/ectools#10]
I've calibrated `ecdaq_case_demo` using the Wattsup Pro meter and got a `sensor.cfg` file.

Then I tried to run `ecdaq_case_demo` with an application but got two strange results.
1) `ecdaq_case_demo` prints an error message when the `sensor.cfg` file is present

``` bash
$ ./ecdaq_case_demo --datafile ectools.csv --logfile ectools.log ./run-vload-script.sh
{ "weights": [6.89938e-310,6.89938e-310], "lr": { "params": 1, "buffer size": 100, "input": [0,0](), "target": [0,0]() } }
DPELinearRegression::operator>>(): <error> Bad input file.
{ "weights": [243,1.34115e-13], "lr": { "params": 1, "buffer size": 100, "input": [0,0](), "target": [0,0]() } }
Data Acquisition ... [start]
Generating load with synchronization pulses
RTS set to low... waiting 5 seconds
Starting
Function terminated.
Start pulse sent at ' 2006-01-04 03:58:01.005114 '; End Pulse sent at ' 2006-01-04 03:59:02.156321 '
Data Acquisition ... [done]
$ 
```

2) the estimation value printed in the CSV file does not change despite the fact that the application is creating random load in the machine. The value remains constant at 243.
I saw in the make file that the tools are compiled against two files from the `libcpufreq`.
One thing I should mention is that the CPU of the machine I'm running the tests in does not support frequency scaling. 

I can send you all the necessary things to reproduce this you you'd like. I just can't attach anything that is not an image to this post :(

Contents of sensor.cfg

```
{ "weights": [243,1.34115e-13], "lr": { "params": 1, "buffer size": 100, "input": [100,2]((1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,0.0870712),(1,0),(1,0),(1,0),(1,0.0025),(1,0),(1,0.00249377),(1,0),(1,0),(1,0),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,0.982544),(1,0.005),(1,0.00249377),(1,0),(1,0),(1,0.0025),(1,0),(1,0),(1,0.00250627),(1,0.0225564),(1,0),(1,0.873057),(1,0.995),(1,0.997506),(1,1),(1,0.995),(1,0.997506),(1,0.997494),(1,0.997494),(1,0.997494),(1,0.997506),(1,0.997494),(1,1),(1,0.997506),(1,0.9975),(1,1),(1,0.995025),(1,1),(1,0.997494),(1,0.9975),(1,0.997494),(1,0.997506),(1,1),(1,0.995025),(1,0.997494),(1,0.9975),(1,1),(1,0.995025),(1,1),(1,0.997481),(1,0.409375),(1,0.00755668),(1,0),(1,0),(1,0.0025),(1,0.00249377),(1,0),(1,0.0025),(1,0),(1,0),(1,0.00249377),(1,0.944584),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1),(1,1)), "target": [100,1]((243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243),(243)) } }
```

Contents of ectools.log

```
#  POWER_DYN_LR_CPU [DPELRCpu]  [ok]
#
#Benchmark summary
#  Command line:        ./run-vload-script.sh
#  Start time:          1136343474
#  End time:            1136343543
```

Partial contents of ectools.csv

```
TS,DPELRCpu
1136343474,243
1136343475,243
1136343476,243
1136343477,243
1136343478,243
1136343479,243
1136343480,243
1136343481,243
1136343482,243
1136343483,243
1136343484,243
1136343485,243
1136343486,243
1136343487,243
1136343488,243
1136343489,243
1136343490,243
1136343491,243
1136343492,243
1136343493,243
1136343494,243
1136343495,243
1136343496,243
1136343497,243
1136343498,243
1136343499,243
1136343500,243
1136343501,243
```

**Comment by cupertino** [comment]
Hello Miguel,

  From the results, it seems that the output from the WattsUp is constant
(243W) while running your script, so the model is calibrated as a constant
to that value. Could you check if the WattsUp power meter is properly
updated over time? Just run a test and collect the data from the wattsUp
while running the 'run-vload-script.sh' script.

**Comment by miguelaferreira** [comment]
Hi Leandro,

While running the script the meter reports variations in the power measurement. However, I only use the meter to calibrate ecdaq, then I expect it to use its own models to estimate the power. Therefore the meter is not even connected to the server where I'm running the tests.

Isn't this the way it is supposed to work?

**Comment by cupertino** [comment]
As described on 'sensor.cfg' file, the target outuput, i.e. the WattsUp
power is always 243, so the data that is used to calibrate is this constant.

The proper calibration/usage should be:
1) calibrate on the machine which contains the WattsUp meter
2) use the 'sensor.cfg' file to run the powwer estimator on any machine
with the same characteristics

**Comment by miguelaferreira** [comment]
The calibration was done on the test server with the WattsUp meter connected.
I will have to check that the calibration was well done or not.
I'll get back to you on that.

What about the message "`DPELinearRegression::operator>>(): <error> Bad input file.`"?

**Comment by cupertino** [comment]
You're right I forgot to talk about it. It seems the used file has no
inputs/targets, i.e. no training data, this should be the error but we
would have to dig a little deeper on it to make sure. It seems you're
collectiong data without a proper configuration file.

**Comment by miguelaferreira** [comment]
>  It seems the used file has no inputs/targets, i.e. no training data, this should be the error but we would have to dig a little deeper on it to make sure.

Which file do you mean by used file?

>  It seems you're collecting data without a proper configuration file.

How do I create such a configuration file?

**Comment by cupertino** [comment]
This is the data from the config file:
{ "weights": [6.89938e-310,6.89938e-310], "lr": { "params": 1, "buffer
size": 100, "input": [0,0](), "target": [0,0]() } }

To create the config file you need to calibrate it:
$ ecdaq_case_demo --calibrate

On this demo we use the stress benchmark to calibrate it but you can set
the bench as you wish on the benchmark function in the demos/ecdaq_case.cpp
file:

void*
benchmark(void\* data)
{
  int ncpus, i = 1;
  char buff[128];

  ncpus = sysconf(_SC_NPROCESSORS_CONF);

  for (i = 1; i <= ncpus; i++)
    {
      sprintf(buff, "stress --cpu %d --timeout 30", i);
      system(buff);
      sleep(10);
      sprintf(buff, "stress --vm %d --timeout 30", i);
      system(buff);
      sleep(10);
      sprintf(buff, "stress --io %d --timeout 30", i);
      system(buff);
      sleep(10);
    }
  sprintf(buff, "stress --cpu %d --vm %d --io %d --timeout 30", ncpus,
ncpus,
      ncpus);
  system(buff);
  sleep(10);

  return (void*) 0;
}

**Comment by miguelaferreira** [comment]
> It seems you're collectiong data without a proper configuration file.
> 
> To create the config file you need to calibrate it:
> `$ ecdaq_case_demo --calibrate`

Calibration (according to the command you mention) was exactly what I did to create the configuration file. But apparently it was not proper.
My guess is that the code used to create the configuration file has a problem.

Thus, the question still remains: how to create a proper configuration file?

**Comment by miguelaferreira** [comment]
Hi Leandro,

After some debugging I figured out why was the WattsUp pro sensor always reporting the same value.
In the calibration process, the update() method of the sensor was only called once (at the start), then afterwards the method getValue() of the sensor was called multiple times.

As you can see from the SensorPowerWattsUp tests, calling getValue() alone will not re-read the value from the sensor.
I see two ways of fixing this:
- Change the way the sensor it used in the calibration process
- Change the getValue() method of the sensor to call update() before returning a new value.

I can fork the repository and create a new pull request with the change, but would like to hear form you about what you think is best.

Cheers

**Comment by cupertino** [comment]
Hello,

  The reason for separating the update() from the getValue() method is
that, for some sensors, we need to update their value once and use it
several times. For instance, this applies for the CPU usage of PIDs. In the
CPU usage, the CPU time sensor is used by updating the total time spend by
the whole system once and, for each requested PID, we get not only the CPU
time of this PID but the total time as well.

  In other words, first case fits better and let its usage more flexible.

Cheers

**Comment by miguelaferreira** [comment]
>  In other words, first case fits better and let its usage more flexible.

I agree with having the flexibility in implementing sensor classes.
But I think that the specifics of the sensor should be encapsulated in the sensor class.
Otherwise, you have to add extra code everywhere you use the sensor to do something like:

```
if(<should call update>) {
    sensor.update();
}
sensor.getValue();
```

Or am I missing something?

**Comment by cupertino** [comment]
At first, we did it like this with the update being called by the getValue,
but we changed it because of the monitoring.

For the CPU usage case we will have: cpu_pid_usage = cpu_pid_time /
cpu_sys_time
If we fetch on different time, we will need to have one cpu_sys_time sensor
for each cpu_pid_usage and we dont want to allocate all this memory, so we
let to the user to decide when to update each sensor.

**Comment by miguelaferreira** [comment]
I see the reasons why in some sensors you don't wanna update every time, but in this case you should. Otherwise you have to use the API in a way for some sensors and in another way for others.

For now you only get one value form the WattsUp Pro meter. Is that how you want the tool to work?
